ortools/sat/timetable_horizontal_edgefinding.cc

```diff
@@ -0,0 +1,290 @@
+
+  // Overload represents the accumulated quantity of energy that could not be
+  // consumed before time.
+  IntegerValue overload = 0;
+
+  // Total demand required at time if all processed tasks were starting at
+  // their start min.
+  IntegerValue demand_req = 0;
+
+  // Total demand required at time if all processed tasks that could overlap
+  // time were. This is used to avoid placing overload in places were no task
+  // would actually be.
+  IntegerValue demand_max = 0;
+
+  while (time < window_end) {
+    // Aggregate the changes of all events happening at time. How to process
+    // an event depends on its type ("full" vs "fixed-part").
+    IntegerValue delta_max = 0;
+    IntegerValue delta_req = 0;
+    while (profile_events_[next_event].time == time) {
+      const ProfileEvent event = profile_events_[next_event];
+      switch (task_types_[event.task_id]) {
+        case TaskType::kIgnore:
+          break;  // drop the event
+        case TaskType::kFull:
+          switch (event.event_type) {
+            case ProfileEventType::kStartMin:
+              delta_max += event.height;
+              delta_req += event.height;
+              break;
+            case ProfileEventType::kEndMin:
+              delta_req -= event.height;
+              break;
+            case ProfileEventType::kEndMax:
+              delta_max -= event.height;
+              break;
+            default:
+              break;
+          }
+          break;
+        case TaskType::kFixedPart:
+          switch (event.event_type) {
+            case ProfileEventType::kStartMax:
+              delta_max += event.height;
+              delta_req += event.height;
+              break;
+            case ProfileEventType::kEndMin:
+              delta_req -= event.height;
+              delta_max -= event.height;
+              break;
+            default:
+              break;
+          }
+          break;
+      }
+      next_event++;
+    }
+
+    // Should always be safe thanks to the sentinel.
+    DCHECK_LT(next_event, profile_events_.size());
+
+    IntegerValue next_time = profile_events_[next_event].time;
+    const IntegerValue length = next_time - time;
+
+    demand_max += delta_max;
+    demand_req += delta_req;
+
+    DCHECK_LE(demand_req, demand_max);
+    DCHECK_GE(overload, 0);
+
+    // The maximum amount of resource that could be consumed if all non-ignored
+    // tasks that could be scheduled at the current time were.
+    const IntegerValue true_capa = std::min(demand_max, capacity);
+
+    // Indicate whether we're using some capacity at this time point. This is
+    // used to decide later on how to update new_window_end.
+    const IntegerValue capa_used = std::min(demand_req + overload, true_capa);
+
+    // Amount of resource available to potentially place some overload from
+    // previous time points.
+    const IntegerValue overload_delta = demand_req - true_capa;
+
+    if (overload_delta > 0) {  // adding overload
```

Comment:
nit: I think this logic would be easier to read if the variable naming made clear what is the overload/removable/used for a time unit and what is the equivalent for the time window.

Reply:
I wasn't super happy with these names either; what do you think about the new ones? I've also slightly simplified the logic as there's no need for the `std::min` call.